de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/SharedResourcesManager.java

```diff
@@ -195,126 +195,194 @@ private void handleFileRecovery(FileActivity activity) throws IOException {
+    private void handleFileDeletion(
+        @NotNull
+            FileActivity activity) throws IOException {
+
+        SPath path = activity.getPath();
+        IFile file = path.getFile();
 
         if (!file.exists()) {
             LOG.warn(
-                "could not delete file " + file + " because it does not exist");
+                "Could not delete file " + file + " as it does not exist.");
+
+            return;
+        }
+
+        if (localEditorHandler.isOpenEditor(path)) {
+            localEditorManipulator.closeEditor(path);
         }
 
         try {
             fileSystemListener.setEnabled(false);
-            FileUtils.delete(file);
-            //HACK: It does not work to disable the fileSystemListener temporarly,
-            //because a fileCreated event will be fired asynchronously,
-            //so we have to add this file to the filter list
-            fileSystemListener
-                .addIncomingFileToFilterFor(file.getLocation().toFile());
+
+            localEditorHandler.saveDocument(path);
+
+            file.delete(DELETION_FLAGS);
+
         } finally {
             fileSystemListener.setEnabled(true);
         }
+
+        annotationManager.removeAnnotations(file);
+
+        //TODO reset the vector time for the deleted file
     }
 
-    private void handleFileCreation(FileActivity activity) throws IOException {
+    private void handleFileCreation(
+        @NotNull
+            FileActivity activity) throws IOException {
 
-        String encodingString = activity.getEncoding() == null ?
-            EncodingProjectManager.getInstance().getDefaultCharset().name() :
-            activity.getEncoding();
+        SPath path = activity.getPath();
+        IFile file = path.getFile();
 
-        //FIXME: Test if updateEncoding method will be necessary
-        String newText = new String(activity.getContent(), encodingString);
+        if (file.exists()) {
+            LOG.warn(
+                "Could not create file " + file + " as it already exists.");
 
-        //if the file exists, try to replace the content completely
-        if (!newText.isEmpty()) {
-            //this is true only when the file already existed
-            boolean replaceSuccessful = localEditorManipulator
-                .replaceText(activity.getPath(), newText);
+            return;
+        }
 
-            if (replaceSuccessful) {
-                //If the content of the existing document was replaced
-                //successfully, save the file
-                localEditorHandler.saveDocument(activity.getPath());
-                return;
-            }
+        InputStream contents = new ByteArrayInputStream(activity.getContent());
+
+        try {
+            fileSystemListener.setEnabled(false);
+
+            file.create(contents, FORCE);
+
+        } finally {
+            fileSystemListener.setEnabled(true);
         }
+    }
+
+    private void handleFolderCreation(
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Move this method into the anonymous class declared at line 105. ([squid:S3398](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3398))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=277&issues=AWagwjc6lHK86HcoyTJD&open=AWagwjc6lHK86HcoyTJD)

[ISSUE_KEY]: AWagwjc6lHK86HcoyTJD